### PR TITLE
flan

### PR DESCRIPTION
**Pull request by nmtadam**
POC to show the basic idea of named pools which are built on top
the root object abstraction discussed previously. Includes a fuse
filesystem of how to build on top of flan

flan = FlexAlloc Named Objects

Signed-off-by: Adam Manzanares <a.manzanares@samsung.com>

### CODE REVIEW

**Comment by Joelgranados** [comment]
Flan is broken with ZNS:
Basically when flan reopens an already existing FlexAlloc state and wants to "update" the root object then it has to re-write it. This means taking a zone that has a write pointer different than 0 and trying to write to lba zero.  This is not allowed in ZNS and fails. I see this when I try to relaunch db_bench on the same Flan instance.

We need to find a way to update the root object without starting in lba zero. log structured approach?

File: flan/flan.c [hunk 1]
```diff
@@ -0,0 +1,883 @@
...
+
+}
+
+struct flan_oinfo *flan_find_oinfo(struct flan_handle *flanh, const char *name)
+{
+  struct flan_oinfo *oinfo;
+
+  flan_reset_pool_dir();
+  while (((oinfo = flan_get_oinfo(flanh, false)) != NULL) &&
+         strncmp(oinfo->name, name, FLAN_OBJ_NAME_LEN_MAX));
```
Comment (Joelgranados) [inline flan/flan.c:272]: I believe that there is an error here.
You use basename for all the names but here you take the full name and compare it with the oinfo->name which has been previously passed through basename. 
1. A temporary fix is to compare the strings in reverse order and only considering the minimum length. This is the solution that I took and it seems to fix the issue of db_bench being able to restart
2. Another, more important questions is: why use basename?

File: flan/flan.c [hunk 2]
```diff
@@ -0,0 +1,883 @@
...
+
+int flan_init_dirhandle(struct flan_handle *flanh)
+{
+  struct fla_object robj;
+  int ret = 0;
+
+  if (flan_dir.cur == FLAN_DHANDLE_NONE)
+  {
+    flan_dir.cur = 0;
+    if (!flanh->is_zns)
```
[the remark this answers is not in the document]
Comment (nmtadam) [reply flan/flan.c:144]: IIRC ZNS keeps consistency differently, no writes to MD until flan exit. CNS would update MD on disk as it operates, we should discuss this further and document. This area needs work.

File: flan/flan.c [hunk 3]
```diff
@@ -0,0 +1,883 @@
...
+  {
+    printf("Error allocating fla object\n");
+    return ret;
+  }
+
+  memcpy(oinfo->name, name, namelen + 1);
+  oinfo->size = 0;
+  oinfo->fla_oh = fla_obj;
+
+  if (!flanh->is_zns)
```
[the remark this answers is not in the document]
Comment (nmtadam) [reply flan/flan.c:247]: This is correct

File: flan/flan.c [hunk 4]
```diff
@@ -0,0 +1,883 @@
...
+  // Cant read past end of file
+  if (offset + len > oinfo->size)
+    len = oinfo->size - offset;
+
+  if (flan_otable[oh].o_flags & FLAN_OPEN_FLAG_WRITE)
+    return flan_object_read_rw(oh, buf, offset, len, flanh, oinfo);
+
+  return flan_object_read_r(oh, buf, offset, len, flanh, oinfo);
+}
+int flan_update_md(struct flan_ohandle *oh, struct flan_handle *flanh)
```
[the remark this answers is not in the document]
Comment (nmtadam) [reply flan/flan.c:632]: Yes, place that needs improvement.

File: flan/flan.c [hunk 5]
```diff
@@ -0,0 +1,883 @@
...
+  {
+    printf("flan_object_write fla object write fails\n");
+    return -EIO;
+  }
+
+  if (offset + len > oinfo->size)
+  {
+    oinfo->size = offset + len;
+
+    if (!flanh->is_zns)
```
[the remark this answers is not in the document]
Comment (nmtadam) [reply flan/flan.c:767]: IIRC yup that is right

**Pull request closed without merging**